mm2src/coins/utxo/slp.rs

```diff
@@ -1,38 +1,603 @@
+    ) -> Result<UtxoTx, String> {
+        let payment_script = payment_script(time_lock, secret_hash, self.platform_utxo.my_public_key(), other_pub);
+        let script_pubkey = ScriptBuilder::build_p2sh(&dhash160(&payment_script)).to_bytes();
+        let slp_out = SlpOutput { amount, script_pubkey };
+        let preimage = try_s!(self.generate_slp_tx_preimage(vec![slp_out]).await);
+        generate_and_send_tx(
+            &self.platform_utxo,
+            preimage.inputs,
+            preimage.outputs,
+            FeePolicy::SendExact,
+            preimage.recently_spent,
+        )
+        .await
+    }
+
+    async fn validate_htlc(
+        &self,
+        tx: &[u8],
+        other_pub: &Public,
+        time_lock: u32,
+        secret_hash: &[u8],
+        amount: BigDecimal,
+    ) -> Result<(), MmError<ValidateHtlcError>> {
+        let mut tx: UtxoTx = deserialize(tx).map_to_mm(ValidateHtlcError::TxParseError)?;
+        tx.tx_hash_algo = self.platform_utxo.as_ref().tx_hash_algo;
+        if tx.outputs.len() < 2 {
+            return MmError::err(ValidateHtlcError::TxLackOfOutputs);
+        }
+
+        let slp_tx: SlpTxDetails =
+            deserialize(tx.outputs[0].script_pubkey.as_slice()).map_to_mm(ValidateHtlcError::OpReturnParseError)?;
+
+        match slp_tx.transaction {
+            SlpTransaction::Send { token_id, amounts } => {
+                if token_id != self.token_id() {
+                    return MmError::err(ValidateHtlcError::InvalidSlpDetails);
+                }
+
+                if amounts.is_empty() {
+                    return MmError::err(ValidateHtlcError::InvalidSlpDetails);
+                }
+
+                let expected = sat_from_big_decimal(&amount, self.decimals())?;
+
+                if amounts[0] != expected {
+                    return MmError::err(ValidateHtlcError::InvalidSlpDetails);
+                }
+            },
+            _ => return MmError::err(ValidateHtlcError::InvalidSlpDetails),
+        }
+
+        let dust_decimal = big_decimal_from_sat_unsigned(self.dust(), self.platform_utxo.decimals());
+        let validate_fut = utxo_common::validate_payment(
+            self.platform_utxo.clone(),
+            tx,
+            1,
+            other_pub,
+            self.platform_utxo.my_public_key(),
+            secret_hash,
+            dust_decimal,
+            time_lock,
+        );
+
+        validate_fut
+            .compat()
+            .await
+            .map_to_mm(ValidateHtlcError::ValidatePaymentError)?;
+
+        Ok(())
+    }
+
+    pub async fn refund_htlc(
+        &self,
+        htlc_tx: &[u8],
+        other_pub: &Public,
+        time_lock: u32,
+        secret_hash: &[u8],
+    ) -> Result<UtxoTx, String> {
+        let tx: UtxoTx = try_s!(deserialize(htlc_tx).map_err(|e| ERRL!("{:?}", e)));
+        let slp_tx: SlpTxDetails =
+            try_s!(deserialize(tx.outputs[0].script_pubkey.as_slice()).map_err(|e| ERRL!("{:?}", e)));
```

Comment:
I guess, it's worth checking `!tx.outputs.is_empty()`. I'm not sure the network wouldn't reject a transaction without outputs.

Reply:
If I'm not mistaken BCH doesn't allow transactions without outputs but I agree that it's worth checking just in case :slightly_smiling_face: